learn/security/tenant_tokens.md

```diff
@@ -0,0 +1,303 @@
+    'filter': `user_id = ${currentUserID}`
+  }
+};
+
+const expiryDate = 1645461882;
+
+const token = client.generateTenantToken(searchRules, expiryDate, apiKey);
+```
+
+As you can see, there are three main elements necessary to generate a tenant token: [search rules](#search-rules), [expiry dates](#expiry-date), and [API keys](#api-keys). You can read more about each one these in [this guide's final section](#tenant-token-settings).
+
+### Using a tenant token with an SDK
+
+The resulting token, stored on the `token` variable, can then be sent back to the front-end and used to make queries that will only return results whose `user_id` attribute equals the current user's ID:
+
+```js
+// placeholder code sample. Final version will be supplied by the integrations team
+// <CodeSamples id="tenant_token_guide_sdk_search_1">
+const frontEndClient = new Meilisearch({
+  apiKey: token
+});
+
+const medicalRecordsIndex = frontEndClient.index('patient_medical_records');
+const search = medicalRecordsIndex.search('blood test');
+```
+
+Applications may use tenant tokens and API keys interchangeably when searching. For example, the same application might use a default search API key for queries on public indexes and a tenant token for logged-in users searching on private data.
+
+## Generating tenant tokens without a Meilisearch SDK
+
+Though [Meilisearch recommends using an official SDK](#generating-tenant-tokens-with-an-sdk), this is not a requirement. Since tenant tokens follow the JWT standard, you can use a number of [compatible third-party helper libraries](https://jwt.io/libraries). You may also skip all assistance and generate a token from scratch.
+
+The full process requires you to [create a token header](#creating-a-token-header), [prepare a data payload](#preparing-the-token-payload) with at least one set of search rules, and then [sign the token](#signing-a-token) with an API key.
+
+::: note
+The following examples use javascript for the sake of simplicity, but Meilisearch should be compatible with the majority of modern stacks and languages.
+:::
+
+### Creating a token header
+
+The first step when generating a tenant token is to create a header stating token type and the chosen encryption algorithm. Once that is done you must encode the header object into `base64`:
+
+```js
+let header = {
+  "alg": "HS256",
+  "typ": "JWT"
+}
+
+const base64Header = base64Encode(header);
+```
+
+The encryption algorithm indicated by `alg` can be one of: `HS256`, `HS384`, or `HS512`. The choice of algorithm has no impact in tenant token behavior.
+
+Meilisearch only supports JWT tokens and `typ` must always be `JWT`.
+
+### Preparing the token payload
+
+Once the header is ready, you must create a payload. In this case, payload means the data containing the token's API key prefix, search rules, and optional expiry date.
+
+First, you must extract the first eight characters from a valid API key with access to the search endpoint:
```

Comment:
Idk if this is super important, but do we need to mention that the `apiKeyPrefix` can't be generated from a master key?